core/envvars.go

```diff
@@ -0,0 +1,179 @@
+package core
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/dagger/dagger/dagql"
+	"github.com/joho/godotenv"
+	"github.com/moby/buildkit/frontend/dockerfile/shell"
+	"github.com/vektah/gqlparser/v2/ast"
+	sh "mvdan.cc/sh/v3/shell"
+)
+
+type EnvVariable struct {
+	Name  string `field:"true" doc:"The environment variable name."`
+	Value string `field:"true" doc:"The environment variable value."`
+}
+
+func (EnvVariable) Type() *ast.Type {
+	return &ast.Type{
+		NamedType: "EnvVariable",
+		NonNull:   true,
+	}
+}
+
+func (EnvVariable) TypeDescription() string {
+	return "An environment variable name and value."
+}
+
+func (EnvVariable) Description() string {
+	return "A simple key value object that represents an environment variable."
+}
+
+// EnvFile represents a collection of environment variables that can be manipulated
+type EnvFile struct {
+	// Variables stored as key-value pairs, preserving order and allowing duplicates
+	Environ []string `json:"variables"`
+	Expand  bool     `json:"expand"`
+}
+
+func (*EnvFile) Type() *ast.Type {
+	return &ast.Type{
+		NamedType: "EnvFile",
+		NonNull:   true,
+	}
+}
+
+func (*EnvFile) TypeDescription() string {
+	return "A collection of environment variables."
+}
+
+// WithVariable adds a new environment variable to the EnvFile
+func (ef *EnvFile) WithVariable(name, value string) *EnvFile {
+	ef = ef.Clone()
+	ef.Environ = AddEnv(ef.Environ, name, value)
+	return ef
+}
+
+// Clone creates a deep copy of the EnvFile
+func (ef *EnvFile) Clone() *EnvFile {
+	cp := *ef
+	cp.Environ = slices.Clone(ef.Environ)
+	return &cp
+}
+
+// WithoutVariable removes all occurrences of the named variable
+func (ef *EnvFile) WithoutVariable(name string) *EnvFile {
+	ef = ef.Clone()
+	ef.Environ = nil
+	for _, kv := range ef.Environ {
+		k, _, _ := strings.Cut(kv, "=")
+		if !shell.EqualEnvKeys(k, name) {
```

Comment:
```suggestion
		if k != name {
```

We don't need the dockerfile shell package here, it's def is:

```go
// EqualEnvKeys compare two strings and returns true if they are equal.
// On Unix this comparison is case-sensitive.
// On Windows this comparison is case-insensitive.
func EqualEnvKeys(from, to string) bool {
	return from == to
}
```

This only runs on linux.